### ISSUE TITLE
Add line number information to JavaClassSource

### ISSUE THREAD

**Issue by orfjackal** [junit-team/junit5#406]
In [Specsy](http://specsy.org/), because the individual tests are not methods but lambdas, it's not possible to use `JavaMethodSource` for IDE navigation. I could dig the line numbers from the lambda classes or by creating a stack trace during test registration, but the only `TestSource` which has line number information is `FileSource` and it's not possible to reliably find out the .java file's path just based on the .class file.

For this purpose JUnit should add `Optional<FilePosition> getPosition()` to `JavaClassSource`.

**Comment by raniejade** [comment]
The same for [Spek](https://github.com/JetBrains/spek).

**Comment by sbrannen** [comment]
If we add `Optional<FilePosition> getPosition()` to `JavaClassSource`, who is going to populate the value?

**Comment by orfjackal** [comment]
The same way as with `FileSource`, i.e. an optional constructor argument.

**Comment by sbrannen** [comment]
> The test engine will populate it the same way as with FileSource, i.e. an optional constructor argument. 

Yes, that's clear. So please allow me to rephrase my question. 😉 

I meant:

How should the `TestEngine` implementation determine the line number of the lambda expression in the Java source code at runtime (especially if the source is not available)?

**Comment by raniejade** [comment]
Been pondering about that actually. One way i can think of is expecting the stacktrace of the current thread. `Thread.currentThread().getStackTrace()`, but it feels a bit hackish for me.

**Comment by orfjackal** [comment]
`Thread.currentThread().getStackTrace()` would be the easiest and most reliable way.

I just now realized that because the lambda classes are generated at runtime, it will not be possible to access their bytecode, unless you do the same tricks as Retrolambda. Thus it will be hard to determine just based on a lambda instance that which private synthetic method inside the test class is the lambda's implementation. Reading the line numbers from the test class bytecode would be easy, but that's not enough if we don't know that which method corresponds a particular lambda instance. Better just stick with `getStackTrace()`.

**Comment by marcphilipp** [comment]
But that will only work when you _execute_ the test method, not when "registering" it, right? Since `TestDescriptors`/`TestIdentifiers` have to be published to listeners (IDEs etc.) before execution that won't help -- or am I missing something?

**Comment by orfjackal** [comment]
In Specsy test registration and execution happen together. Or to be more exact, tests are registered during the first execution of their enclosing test. See the Java example on http://specsy.org/ front page. Each call to the `spec` method registers a test (dynamically) and then executes the test (or skips it due to the test isolation algorithm).

**Comment by raniejade** [comment]
In Spek we pass around callbacks to build the tests, we _execute_ this callbacks during the discovery to build the test tree. We are taking advantage of several language features, to make it feel more like a built-in language construct.


This
```kotlin
class SomeSpec: Spek({
    describe("some foo") {
       it("should pass") { .. }
    }
})
```

is equivalent to
```kotlin
class SomeSpec: Spek({
    describe("some foo", {
       it("should pass", { .. })
    })
})
```

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
This has been implemented in `master` in commit a381e6ea5567c3abc3cdc4da40951bf9fac56c8f.

Please let us know if this fulfills your needs.
